### PR TITLE
Collapse duplicate moveShape mutations server-side.

### PR DESCRIPTION
**Pull request by aboodman**
This is suprisingly more sublte than it at first seems like it would be.

### CODE REVIEW

File: backend/user_error.ts [hunk 1]
```diff
@@ -1,5 +0,0 @@
-export class UserError extends Error {
```
Comment (aboodman) [inline backend/user_error.ts:1]: I removed the concept of `UserError`.

Working with these backends is already subtle and having this extra error mode makes it more so.

This means that the system is going to bias toward failing in the stuck direction, which also isn't good.

I'm not sure what the solution is, but opting for simplicity for now.

File: pages/api/replicache-push.ts [hunk 2]
```diff
@@ -1,66 +1,77 @@
...
-      if (mutation.id < expectedMutationID) {
-        console.log('This mutation has already been processed. Nothing to do.');
-        return;
-      }
-      if (mutation.id > expectedMutationID) {
-        console.log('This mutation is from the future. Nothing to do but wait.');
-        return;
-      }
+      // Scan forward from here collapsing any collapsable mutations.
+      for (let mutation: Mutation; mutation = push.mutations[i]; i++) {
```
Comment (arv) [inline pages/api/replicache-push.ts:40]: Reading out of bounds deopts. Please check the length

File: pages/api/replicache-push.ts [hunk 3]
```diff
@@ -78,6 +89,16 @@ export default async (req: NextApiRequest, res: NextApiResponse) => {
   res.status(200).json({});
 };
 
+// If prev and next are collapsible, collapse them by mutating next.
+function collapse(prev: Mutation, next: Mutation): boolean {
+  if (prev.name == 'moveShape' && next.name == 'moveShape') {
```
Comment (arv) [inline pages/api/replicache-push.ts:94]: `===` -- get into the habit

**Review by arv — COMMENTED** [review]
LGTM

**Pull request merged**